src/jdk.incubator.code/share/classes/jdk/incubator/code/parser/impl/DescParser.java

```diff
@@ -51,190 +46,111 @@ public static TypeElement.ExternalizedTypeElement parseExTypeElem(String desc) {
-        return parseMethodRef(s);
-    }
-
-    /**
-     * Parse a constructor reference from its serialized textual form.
-     *
-     * @param desc the serialized constructor reference
-     * @return the constructor reference
-     */
-    public static ConstructorRef parseConstructorRef(String desc) {
-        Scanner s = Scanner.factory().newScanner(desc);
-        s.nextToken();
-        return parseConstructorRef(s);
-    }
-
-    /**
-     * Parse a field reference from its serialized textual form.
-     *
-     * @param desc the serialized field reference
-     * @return the field reference
-     */
-    public static FieldRef parseFieldRef(String desc) {
-        Scanner s = Scanner.factory().newScanner(desc);
-        s.nextToken();
-        return parseFieldRef(s);
-    }
-
-    /**
-     * Parse a record type reference from its serialized textual form.
-     *
-     * @param desc the serialized record type reference
-     * @return the record type reference
-     */
-    public static RecordTypeRef parseRecordTypeRef(String desc) {
-        Scanner s = Scanner.factory().newScanner(desc);
-        s.nextToken();
-        return parseRecordTypeRef(s);
-    }
-
-    public static TypeElement.ExternalizedTypeElement parseExTypeElem(Lexer l) {
-        StringBuilder identifier = new StringBuilder();
-        if (l.token().kind == TokenKind.HASH || l.token().kind == TokenKind.AMP) {
-            // Quoted identifier
-            Token t = l.token();
-            while (t.kind != TokenKind.LT) {
-                identifier.append(t.kind == TokenKind.IDENTIFIER ? t.name() : t.kind.name);
-                l.nextToken();
-                t = l.token();
-            }
-        } else {
-            // type element identifier
-            Tokens.Token t = l.accept(TokenKind.IDENTIFIER,
-                    TokenKind.PLUS, TokenKind.SUB, TokenKind.DOT);
-            identifier.append(t.kind == TokenKind.IDENTIFIER ? t.name() : t.kind.name);
-            if (t.kind == TokenKind.IDENTIFIER) {
-                // keep going if we see "."
-                while (l.acceptIf(Tokens.TokenKind.DOT)) {
-                    identifier.append(Tokens.TokenKind.DOT.name);
-                    t = l.accept(Tokens.TokenKind.IDENTIFIER);
-                    identifier.append(t.name());
-                }
-            }
+    //    ExType:
+    //        ExType ExIdent ExType             // infix type application
+    //        ExIdent ExTypeArgs                // prefix type application
+    //        ExTypeArgs                        // anonymous prefix type application
+    //
+    //    ExTypeArgs
+    //        '<' ExType* '>'
+    //        '(' ExType* ')'
+    //        '[' ExType* ']'
+    //        '{' ExType* '}'
+    //
+    //    ExIdent:
+    //        ident
+    //        ident '.' ident
+    //        StringLiteral
+    //        ExSymbol
+    //
+    //    ExSymbol:
+    //        '::'
+    //        ':'
+    //        '?'
```

Comment:
I initially listed `->` here -- but I quickly ran into an issue where parsing a func op would fail, because `OpWriter` also uses the `->` token to separate the return type of the func op from the body (e.g. the ex type parser for the return type of the func op ends up consuming tokens that are not meant for it).